README.md
twitterapiR
================

<!-- badges: start -->

[![R-CMD-check](https://github.com/tangaot/twitterapiR/workflows/R-CMD-check/badge.svg)](https://github.com/tangaot/twitterapiR/actions)
<!-- badges: end -->

## Description

This project is an API wrapper for Twitter in R. This package provides
user friendly functions for the users to query information from Twitter.

The current list of functions included in the wrapper are:

-   set_bearer(“API key”, “API secret”)
-   get_bearer()
-   user_friends(screen_name, number)
-   searchTweets(searchString, resultType, count)
-   followersCount(screen_name)

## Installation

``` r
if (!require("remotes")) {
  install.packages("remotes")
}
remotes::install_github("tangaot/twitterapiR")
```

## Authentication

You will need to follow these instructions to continue. We are accessing
***Twitter*** programatically. ***twitterapiR*** uses the **httr**
package under the hood to manage this.  
The first step is to create a ***Twitter*** application for yourself. To
do so, go to the [Developer Platform’s
page](https://developer.twitter.com/en/docs/twitter-api/getting-started/getting-access-to-the-twitter-api)
and log in.

Follow the instruction and fill in some basic information. After your project
is created, you can generate your consumer API and Secret key.

In your R session, you will want to do the following using the
appropriate values from the web page:

                                set_bearer("API key", "API secret")
                                

This will authenticate via **httr**, we recommend looking at the Token
man page of this package for more information regarding how to manage
the authentication and caching processes.

## Getting Started

This document is intended to describe the usage of each function and to
show examples of each function. To explore the source code or report
some issues, we recommend going to our
[GitHub](https://https://github.com/tangaot/twitterapiR). Before
exploring our functionality, please make sure you complete the following
steps.

## Function Usage

The `set_bearer()` function sets key and secret as environment
variables.

-   **key**: Twitter API key is used to authenticate requests.
-   **secret**: Twitter API secret is used to authenticate
    requests.

``` r
set_bearer('YOUR-KEY-HERE', 'YOUR-SECRET-HERE')
```

The `get_bearer` function is a helper function that can be used to
generate the bearer key by using `API key` and `API secret`. It is
embedded in the functions `searchTweets()`, `user_friends()`, and
`followersCount()`. The bearer key allows the users to query information
from Twitter.

``` r
get_bearer()
```

The `searchTweets()` function can be used to search for related tweets
that match the specified string.

Parameters:

-   **searchString**: a string of a maximum of 100 characters for the desired
    search characters in a tweet. It can contain hashtags in order to search
    specified tags in Twitter.

-   **count**: count will restrict the maximum number of tweets to return.
    The default is 5.

-   **resultType**: specifies the kind of search results you
    want to receive. The default is “mixed.” Valid values
    include:

                      - mixed: returns both popular and real time results in the response

                      - recent: returns only the most recent results in the response

                      - popular: returns only the most popular results in the response

The function will return a DataFrame that contains the following columns:

-   created_time: the date and time the tweet was created
-   user_name: the name of the user that posted the tweet
-   user_screen_name: the screen name of the user that posted the tweet
-   user_followers_count: the number of followers that (re)posted the
    tweet
-   text: the text within the tweet
-   truncated: indicates if the tweet is truncated
-   favorited: indicates if the tweet was favorited
-   retweeted: indicates it the tweet was retweeted
-   favorite_count: the number of times the tweet was favorited
-   retweet_count: the number of times the tweet was retweeted

In the example below `searchTweets()` function returns all the users who
recently used #ubc.

``` r
# Use `searchTweets()` function to search for the top 100 recent tweets using #ubc
ubc_tweets <- searchTweets("#ubc", resultType = "recent",count = 100)

head(ubc_tweets)
#>                     created_time     user_name user_screen_name
#> 1 Mon Feb 21 00:03:22 +0000 2022   Park Ranger        netminnow
#> 2 Sun Feb 20 23:45:04 +0000 2022         Sandy         sandysva
#> 3 Sun Feb 20 23:01:07 +0000 2022     16forward        16forward
#> 4 Sun Feb 20 22:17:50 +0000 2022 Richard Ajabu     richardajabu
#> 5 Sun Feb 20 20:52:01 +0000 2022   Toronto_Bot      Toronto_Bot
#> 6 Sun Feb 20 20:48:22 +0000 2022 Richard Ajabu     richardajabu
#>   user_followers_count
#> 1                 5349
#> 2                  833
#> 3                   43
#> 4                    0
#> 5                  401
#> 6                    0
#>                                                                                                                                            text
#> 1 RT @richardajabu: @mccaffreyr3 @selectedwisdom As Putin continues the slow motion conversion of EUkraine to RUkraine,\nwhy isn't the west re…
#> 2 RT @richardajabu: @mccaffreyr3 @selectedwisdom As Putin continues the slow motion conversion of EUkraine to RUkraine,\nwhy isn't the west re…
#> 3 RT @richardajabu: @mccaffreyr3 @selectedwisdom As Putin continues the slow motion conversion of EUkraine to RUkraine,\nwhy isn't the west re…
#> 4   Complete Fail. Sad for UK. 🙄\n\nhttps://t.co/sRps0Vf2HP\n#cdnpoli #uspolitics #ukpolitics #eu #auspol #nzpol #yow… https://t.co/V4HroUXrAq
#> 5 RT @richardajabu: @mccaffreyr3 @selectedwisdom As Putin continues the slow motion conversion of EUkraine to RUkraine,\nwhy isn't the west re…
#> 6 @mccaffreyr3 @selectedwisdom As Putin continues the slow motion conversion of EUkraine to RUkraine,\nwhy isn't the w… https://t.co/IofYO2JrWD
#>   truncated favorited retweeted favorite_count retweet_count
#> 1     FALSE     FALSE     FALSE              0             4
#> 2     FALSE     FALSE     FALSE              0             4
#> 3     FALSE     FALSE     FALSE              0             4
#> 4      TRUE     FALSE     FALSE              0             0
#> 5     FALSE     FALSE     FALSE              0             4
#> 6      TRUE     FALSE     FALSE              8             4
```

``` r
# Select those with start=5 and end time=13
created_time <- substr(ubc_tweets$created_time, start = 5, stop = 13)

# Format the time to month, day, and hour
created_time <- strptime(created_time,format='%b %d %H')

# Save the results in a DataFrame
created_time_df <- as.data.frame(created_time)

created_time_df <-  created_time_df %>% group_by(created_time) %>% summarize(num_tweets = n())

created_time_df
#> # A tibble: 36 × 2
#>    created_time        num_tweets
#>    <dttm>                   <int>
#>  1 2022-02-18 19:00:00          1
#>  2 2022-02-18 20:00:00          9
#>  3 2022-02-18 21:00:00         10
#>  4 2022-02-18 22:00:00          1
#>  5 2022-02-18 23:00:00          6
#>  6 2022-02-19 00:00:00          4
#>  7 2022-02-19 01:00:00          3
#>  8 2022-02-19 02:00:00          2
#>  9 2022-02-19 03:00:00          1
#> 10 2022-02-19 06:00:00          2
#> # … with 26 more rows
```

``` r
# Plot the results
ggplot(created_time_df, aes(x = created_time, y = num_tweets)) + 
    geom_area(fill="firebrick") +
    geom_line(color="black", size=1) +
    geom_point(color="black", size=3) +
    theme_bw()+
    labs(x = "Created Time", y= "Number of Tweets")+
    ggtitle("The Latest 100 Trending Tweets Containing the Hashtag #ubc")
```

<img src="figures/README-unnamed-chunk-7-1.png" width="100%" />

``` r
    theme(panel.grid.major.x = element_blank(),panel.grid.minor.x = element_blank())
#> List of 2
#>  $ panel.grid.major.x: list()
#>   ..- attr(*, "class")= chr [1:2] "element_blank" "element"
#>  $ panel.grid.minor.x: list()
#>   ..- attr(*, "class")= chr [1:2] "element_blank" "element"
#>  - attr(*, "class")= chr [1:2] "theme" "gg"
#>  - attr(*, "complete")= logi FALSE
#>  - attr(*, "validate")= logi TRUE
```

The `followersCount()` function can be used to quickly get the number of
followers a user has by using the user’s screen name.

Parameters:

-   **screen_name**: A string of a Twitter screen name.

In the plot below we use this function in a for loop to get the
followers of the following state leaders as of 2022:

-   Joe Biden -President of the United States
-   Justin Trudeau -Prime Minister of Canada
-   Boris Johnson -Prime Mnister of the United Kingdom
-   Emmanuel Macron -President of France
-   Moon Jae-in -President of South Korea

``` r
# Use followersCount to find the number of follower for each country leader
list = c("POTUS", "JustinTrudeau", "BorisJohnson", "EmmanuelMacron", "moonriver365")
followers = c()
for (i in list){
    followers[i]=followersCount(screen_name =i)$followers_count
    Sys.sleep(1)
}

# Save the results in a DataFrame
followers <- as.data.frame(followers)

# Add name and no_of_followers as column names
followers["name"] <- c("Joe Biden", "Justin Trudeau", "Boris Johnson", "Emmanuel Macron", "Moon Jae-in" )
followers["no_of_followers"] <- followers$followers

# Remove redundant column
followers$followers = NULL

followers
#>                           name no_of_followers
#> POTUS                Joe Biden        18144191
#> JustinTrudeau   Justin Trudeau         5942556
#> BorisJohnson     Boris Johnson         3811533
#> EmmanuelMacron Emmanuel Macron         7633738
#> moonriver365       Moon Jae-in         1952983
```

If you want to visualize the comparison, you can plot the results as
follows.

``` r
# pre-set the bw theme
theme_set(theme_bw())  

# Plot total number of followers per president
g <- ggplot(followers, aes(name, no_of_followers))

# Use a barplot
g + geom_col() + 
  labs(subtitle="example plot", 
       y="number of followers", 
       x="name", 
       title="Popular President?", 
       )
```

<img src="figures/README-unnamed-chunk-9-1.png" width="100%" />

The `user_friends()` function can be used to retrieve the number of
friends a user has by using the user’s screen name. By default the
maximum number of returned friends is 195.

Parameters:

-   **screen_name**: A string of one Twitter screen name.
-   **number**: The number of friends to output. The default is 195.

Similar to the above plot, the plot below uses the `user_friends()`
function in a for loop to get the friends of the following state leaders
as of 2022:

-   Joe Biden - President of the United States
-   Justin Trudeau - Prime Minister of Canada
-   Boris Johnson - Prime Minister of the United Kingdom
-   Emmanuel Macron - President of France
-   Moon Jae-in - President of South Korea

``` r
# Use user_friends to find the number of friends for each country leader
list = c("POTUS", "JustinTrudeau", "BorisJohnson", "EmmanuelMacron", "moonriver365")
friends = c()
for (i in list){
    friends[i]=nrow(user_friends(screen_name=i))
    Sys.sleep(1)
}

# Store the results in a DataFrame
friends <- as.data.frame(friends)

# Add name and no_of_friends as column names
friends["name"] <- c("Joe Biden", "Justin Trudeau", "Boris Johnson", "Emmanuel Macron", "Moon Jae-in" )

friends["no_of_friends"] <- friends$friends

# Remove redundant column
friends$friends = NULL

friends
#>                           name no_of_friends
#> POTUS                Joe Biden            12
#> JustinTrudeau   Justin Trudeau           195
#> BorisJohnson     Boris Johnson           195
#> EmmanuelMacron Emmanuel Macron           195
#> moonriver365       Moon Jae-in           195
```

If you want to visualize the comparison, you can plot the results as
follows.

``` r
# pre-set the bw theme
theme_set(theme_bw())

# Plot the results
g <- ggplot(friends, aes(name, no_of_friends))

# Add barplot
g + geom_col() + 
  labs(subtitle="example plot", 
       y="number of friends", 
       x="name", 
       title="President with the Most friends?", 
       )
```

<img src="figures/README-unnamed-chunk-11-1.png" width="100%" />

## Code of Conduct

Please note that the twitterapiR project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# twitterapiR 0.1.0

---

### NEWS.md setup

- added NEWS.md creation with newsmd

test/testthat.R
library(testthat)
library(twitterapiR)

test_check("twitterapiR")

test/test_get_bearer.R
library(testthat)
#library(twitterapiR)

test_that("API key is retreived", {
    Sys.setenv(key = 'Tds7CCMzPsJbRZH7aikpKiObN')
    Sys.setenv(secret = 'lPH7pIQIS1pVinxh48xvXWgqzGa9gre4Utb9tIZ2W1U0nSCgrz')
    expect_type(get_bearer(), "list")
})

test_that("Invalid bearer key correctly fails", {
    # invalid key and secret
    Sys.setenv(key = 'xxxxxxxxxxxxxxxxx')
    Sys.setenv(secret = 'xxxxxxxxxxxxxxxxxxx')
    expect_error(get_bearer())
    # invalid key and secret
    Sys.unsetenv("key")
    Sys.unsetenv("secret")
    expect_error(get_bearer())
})


test/test_user_friends.R
library(testthat)

Sys.setenv(key = 'Tds7CCMzPsJbRZH7aikpKiObN')
Sys.setenv(secret = 'lPH7pIQIS1pVinxh48xvXWgqzGa9gre4Utb9tIZ2W1U0nSCgrz')

test_that("Incorrect inputs", {
    # invalid screen_name
    expect_error(user_friends(screen_name="BarackObamaa", number =10))
    # invalid number
    expect_error(user_friends(screen_name="BarackObama", number = 2000))
})

test_that("output is a data frame",{
    output <- user_friends(screen_name="BarackObama", number =10)
    expect_true(is.data.frame(output))
})



R/get_bearer.R
get_bearer <- function() {
     #' get_bearer
     #' 
     #' @description This function is a helper function that can be used to help to generate the bearer key using API key and API secret key.
     #' 
     #' @usage get_bearer()
     #' @return If the key and secret are valid, the function will return a list of the type and key.
     #' 
     #' @details The key and secret are required to use in the functions within the twitterapiR package in order to query information from Twitter.
     #' 
     #' @examples
     #' key = 'YOUR-KEY-HERE'
     #' secret = 'YOUR-SECRET-HERE'
     #' get_bearer()
     #' 
     #' @export
     #' 
     
    # get the key and secret from the system environment
    key <- Sys.getenv("key")
    secret<-Sys.getenv("secret")
     
    # raise an error if there are no environment variables named "key" or "secret"
    if (key == "" | secret == "") {
        stop("You must provide an API key to the function or use `Sys.setenv(key = 'your_key_here')`, `Sys.setenv(secret = 'your_secret_key_here')`")
    }
     
    # concatenate key and secret as a bearer key.
    app_keys <- openssl::base64_encode(paste0(key, ":", secret))
    r <- httr::POST("https://api.twitter.com/oauth2/token",
                    httr::add_headers(Authorization = paste0("Basic ", app_keys)),
                    body = list(grant_type = "client_credentials"))
    bearer <- httr::content(r, encoding = "UTF-8")
    
     # raise an error if the bearer key is invalid. Otherwise return the bearer key as a list.
    if (!is.null(bearer$errors)){
        stop("Invalid key or secret, please see the following details: https://developer.twitter.com/en/docs/twitter-api/getting-started/getting-access-to-the-twitter-api")
    }else{
        return(bearer)
    }
}


R/set_bearer.R
set_bearer <- function(key, secret){
    
    #' set_bearer
    #' 
    #' @description This function will set key, and secret as environment variables.
    #' 
    #' @param key character. Twitter API key is used to authenticate requests.
    #' @param secret character. Twitter API secret is used to authenticate requests.
    #' 
    #' @usage set_bearer(key, secret)
    #' @return If the key and secret are successfully set, the function will return TRUE, otherwise, return FALSE.
    #' 
    #' @details The key and secret are always required, which are essentially the username and password for your API. 
    #' To generate this key do so via https://developer.twitter.com/
    #' 
    #' @examples
    #' key = 'YOUR-KEY-HERE'
    #' secret = 'YOUR-SECRET-HERE'
    #' result <- set_bearer(key, secret) 
    #' print(result)
    #' 
    #' @export
    #' 
    
    if (key == "" || secret == "" ) {
        stop("You must provide an API key to the function!!")
    }
    
    #set key and secret as environment variable
    Sys.setenv("key"= key) 
    Sys.setenv("secret" = secret)
    
    tryCatch(
        expr = {
            my_key <- Sys.getenv("key")
            my_secret<-Sys.getenv("secret")
        },
        error = function(e){
            message('Can not set key and secret!') #if we can not get key and secret
            print(e)
            return(FALSE)
        },
        finally = {
            if (my_key == key && my_secret == secret){
                return(TRUE)
            } else {  #if key or secret are not match
                message('Error in set key and secret!')
                return(FALSE)
                
            }
        }
    )    
}

R/user_friends.R
user_friends <- function(screen_name, number = 195){
     #' user_friends
     #' 
     #' @description This function will accept a screen name of a user and output the user's name and screen name as well as the names of the user's friends.
     #' 
     #' @param screen_name character. A string of a screen name on Twitter. 
     #' @param number integer. The number of friends to output. Default is 195.
     #' 
     #' 
     #' @usage user_friends(screen_name, number = 195)
     #' @return A DataFrame contains following columns:
     #' name: The name of the user's friends
     #' screen_name: The screen name of the friends
     #' description: The description of the friends
     #' 
     #' @details This will only work correctly with users that have public profiles, or if you are authenticated and granted access.
     #' 
     #' The screen_name is always required and the profile must be public.
     #' The default number of returned followers is up to a maximum of 195.
     #' 
     #' @references <https://developer.twitter.com/en/docs/twitter-api/v1/accounts-and-users/follow-search-get-users/api-reference/get-friends-list>
     #' @examples
     #' user_friends(screen_name ="BarackObama", 10)
     #' 
     #' @export
     #' 
    
    # raise an error if the maximum number of user is more than 195.
    if (number > 195){
        stop("The number of users to return per page, up to a maximum of 195")
    }
    
    # get the bearer key
    bearer = get_bearer()
    
    base_url = "https://api.twitter.com/1.1/friends/list.json?"
    
    full_url = glue(base_url, 'screen_name={screen_name}&count={number}')
    
    # retrieve information from Twitter
    res=GET(full_url, add_headers(
        Authorization=paste0("Bearer ", bearer$access_token))
    )
    
    #check connection
    if (res$status_code != '200'){
        stop(paste("Connection Failed! Please see the following detail: \n",res))
    }
    
    obj <- httr::content(res, as = "text")
    
    json_data <- fromJSON(obj, flatten = TRUE)
    
    name <- NULL
    description <- NULL
    
    # output the 3 columns as a data frame
    data <- json_data$users  %>% select(name,screen_name, description) %>% as.data.frame()
    
    return(data)
}


R/followersCount.R
followersCount <- function(screen_name){
    #' followersCount
    #' 
    #' @description This function will accept the screen name of a user and output the total number of followers of that user
    #' 
    #' @param screen_name character. A string of one Twitter screen name. 
    #' 
    #' @usage followersCount(screen_name)
    #' @return A DataFrame containing the following columns:
    #' screen_name: The screen name of a user
    #' followers_count: The number of followers of a user
    #' 
    #' @references <https://developer.twitter.com/en/docs/twitter-api/v1/accounts-and-users/follow-search-get-users/api-reference/get-followers-list>
    #' @examples
    #' followersCount(screen_name ="BarackObama")
    #' 
    #' @export
    #' 
    
    # raise an error if the number of characters of the inputted screen name is greater than 100 and if the number of characters of the inputted screen name is 0.
    if (nchar(screen_name) > 100) {
        stop("screen_name can only be up to 100 characters")
    } else if (nchar(screen_name) == 0) {
        stop("screen_name cannot be empty")
    }
    
    # get the bearer key
    bearer = get_bearer()

    # retrieve information from Twitter
    screen_name <- utils::URLencode(screen_name, reserved = TRUE) # encode characters in URLs
    base <- "https://api.twitter.com/1.1/users/show.json?screen_name="
    url <- paste0(base, screen_name)
    res = httr::GET(url, httr::add_headers(Authorization=paste0("Bearer ", bearer$access_token)))
    
    # check connection
    if (res$status_code != '200'){
    stop(paste("Connection Failed! Please see the following detail: \n",res))
    }
    
    obj <- httr::content(res, as = "text")
    
    json_data <- jsonlite::fromJSON(obj, flatten = TRUE)
    
    # ouputs the screen_name and followers count as a data frame.
    screen_name <- as.data.frame(json_data$screen_name)
    followers_count <- as.data.frame(json_data$followers_count)
    data <- merge(screen_name, followers_count)
    colnames(data) <- c('screen_name', 'followers_count')
    

    return(data)
    
}



R/followersCount function.R
followersCount <- function(key, secret, screen_name){
    
    if (nchar(screen_name) > 100) {
        stop("screen_name can only be up to 100 characters")
    } else if (nchar(screen_name) == 0) {
        stop("screen_name cannot be empty")
    }
    
    
    app_keys <- openssl::base64_encode(paste0(key, ":", secret))
    r <- httr::POST("https://api.twitter.com/oauth2/token",
                httr::add_headers(Authorization = paste0("Basic ", app_keys)),
                body = list(grant_type = "client_credentials"))
    bearer <- httr::content(r, encoding = "UTF-8")

    screen_name <- URLencode(screen_name, reserved = TRUE) # encode characters in URLs

    base <- "https://api.twitter.com/1.1/users/show.json?screen_name="

    url <- paste0(base, screen_name)

    res = httr::GET(url, httr::add_headers(Authorization=paste0("Bearer ", bearer$access_token)))
    
    #check connection
    if (res$status_code != '200'){
    stop(paste("Connection Failed! Please see the following detail: \n",res))
    }
    
    obj <- httr::content(res, as = "text")
    
    json_data <- jsonlite::fromJSON(obj, flatten = TRUE)
    screen_name <- as.data.frame(json_data$screen_name)
    followers_count <- as.data.frame(json_data$followers_count)
    data <- merge(screen_name, followers_count)
    colnames(data) <- c('screen_name', 'followers_count')
    

    return(data)
    
}
try_obama <- followersCount('Tds7CCMzPsJbRZH7aikpKiObN', 'lPH7pIQIS1pVinxh48xvXWgqzGa9gre4Utb9tIZ2W1U0nSCgrz', 'BarackObama')
try_obama


R/searchTweets.R
searchTweets <- function(searchString, resultType = "mixed", count = 5){

    
    #' searchTweets
    #' 
    #' @description This function will search for related tweets that match the specified string.
    #' 
    #' @param searchString character. A search string of 100 characters maximum, including operators.
    #' @param resultType character. If not NULL, returns filtered tweets as per value. See details for allowed values.
    #' @param count integer. If not NULL, restricts the maximum number of tweets return. 
    #' Default is 5, up to a maximum of 100
    #' 
    #' 
    #' @usage searchTweets(searchString, resultType = "mixed", count = 5)
    #' @return A DataFrame that contains the following columns:
    #' created_time: the time the tweet was created
    #' user_name: the name of the user
    #' user_screen_name: the screen name of the user
    #' user_followers_count: the number of followers of the user
    #' text: the text of the tweet
    #' truncated: if the tweet is truncated
    #' favorited: if the tweet was favorited
    #' retweeted: if the tweet was retweeted
    #' favorite_count: the number of times the tweet was favorited
    #' retweet_count: the number of times the tweet was retweeted
    #' 
    #' @details This function will return any authorized tweet that matches the search conditions. 
    #' Authorized tweets are public tweets as well as protected tweets that are authenticated by 
    #' registerTwitterOAuth, which users can use. Please note that Twitter's search service and, 
    #' by extension, the Search API is not meant to be an exhaustive source of Tweets.
    #' 
    #' The searchString is always required. It can contain hashtags in order to search specific tags in Twitter.
    #' The resultType specifies the type of search results that you want. 
    #' The current default is "mixed." Valid values include:
    #' mixed : includes both popular and real time results in the response
    #' recent : returns only the most recent results in the response
    #' popular : returns only the most popular results in the response
    #' 
    #' @references <https://developer.twitter.com/en/docs/twitter-api/v1/tweets/search/api-reference/get-search-tweets>
    #' @examples
    #' ##search for hashtag #canada
    #' searchTweets('#canada', resultType = "popular", count = 15)
    #' 
    #' @export
    #' 
    
    
    #check searchString
    if (nchar(searchString) > 100) {
        stop("searchString can only be up to 100 characters!")
    } else if (nchar(searchString) == 0) {
        stop("searchString can not be empty!")
    }
    
    #check count
    count <- as.integer(count)
    if (count <= 0) {
        stop("Count must be positive!")
    } else if(count > 100){
        stop("Count must less than 100!")
    }
    
    #check resultType
    if (!(resultType %in% c("mixed", "recent", "popular"))){
        stop("resultType can only be mixed or recent or popular!")
    }
    
    # get key and secret
    bearer <- get_bearer()
    
    searchString <- utils::URLencode(searchString, reserved = TRUE) # encode characters in URLs
    
    base <- "https://api.twitter.com/1.1/search/tweets.json?q="
    
    url <- paste0(base,searchString,"&count=",count,"&resultType=",resultType)
    
    #get from API
    res = httr::GET(url, httr::add_headers(Authorization=paste0("Bearer ", bearer$access_token)))
    
    #check connection
    if (res$status_code != '200'){
        stop(paste("Connection Failed! Please see the following detail: \n",res))
    }
    
    obj <- httr::content(res, as = "text")
    
    json_data <- jsonlite::fromJSON(obj, flatten = TRUE) #convert to json
    
    data <- as.data.frame(json_data) #convert json to dataframe
    
    statuses.created_at <- NULL
    statuses.user.name  <- NULL
    statuses.user.screen_name  <- NULL
    statuses.user.followers_count <- NULL
    statuses.text <- NULL
    statuses.truncated <- NULL
    statuses.favorited <- NULL
    statuses.retweeted <- NULL
    statuses.favorite_count <- NULL
    statuses.retweet_count <- NULL
    
    #select columns that we are interested
    data <- dplyr::select(data, 
                          statuses.created_at,
                          statuses.user.name, 
                          statuses.user.screen_name,
                          statuses.user.followers_count,
                          statuses.text, 
                          statuses.truncated, 
                          statuses.favorited,
                          statuses.retweeted,
                          statuses.favorite_count,
                          statuses.retweet_count)
    
    data <- dplyr::rename(data,
                          created_time = statuses.created_at,
                          user_name = statuses.user.name,
                          user_screen_name = statuses.user.screen_name,
                          user_followers_count = statuses.user.followers_count,
                          text = statuses.text,
                          truncated = statuses.truncated,
                          favorited = statuses.favorited,
                          retweeted = statuses.retweeted,
                          favorite_count = statuses.favorite_count,
                          retweet_count = statuses.retweet_count)
    
    return(data)
}